replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/Worker.java

```diff
@@ -0,0 +1,335 @@
+   * @param siteManager site manager for accessing sites
+   * @param nodeAdapters node adapter factory
+   */
+  Worker(Queue queue, NodeAdapter local, SiteManager siteManager, NodeAdapters nodeAdapters) {
+    this.queue = queue;
+    this.local = local;
+    this.siteManager = siteManager;
+    this.nodeAdapters = nodeAdapters;
+  }
+
+  /**
+   * Begin monitoring a queue for tasks. The handler of this worker is responsible for halting this
+   * thread by calling interrupt on this worker's thread, finishing the transfer of the current task
+   * if necessary.
+   */
+  @Override
+  public void run() {
+    while (running) {
+      Task task = null;
+      try {
+        task = queue.take();
+        this.processing = true;
+        if (supported(task)) {
+          this.handle(task);
+        } else {
+          task.unlock();
+        }
+        this.processing = false;
+      } catch (InterruptedException e) {
+        if (task != null) {
+          try {
+            task.unlock();
+          } catch (InterruptedException ignore) { // fallthrough
+          }
+        }
+        interrupt();
+        break;
+      } finally {
+        try {
+          local.close();
+        } catch (IOException e) {
+          LOGGER.debug("Failed to close adapter", e);
+        }
+      }
+    }
+  }
+
+  /**
+   * Set this workers running flag to false if it is not currently transferring an item
+   *
+   * @return {@code true} if the worker was cancelled, otherwise {@code false}
+   */
+  boolean cancelIfProcessing() {
+    if (processing) {
+      this.running = false;
+      return true;
+    }
+    return false;
+  }
+
+  private boolean supported(Task task) {
+    if (!OperationType.UNKNOWN.equals(task.getOperation())) {
+      return false;
+    }
+
+    List<MetadataInfo> infos = task.metadatas().collect(Collectors.toList());
+    if (infos.size() > 1) {
+      LOGGER.warn("More than one metadata is currently not supported by this worker");
+      return false;
+    }
+
+    DDFMetadataInfo ddfInfo = infos.stream().filter(DDFMetadataInfo.class::isInstance)
+        .map(DDFMetadataInfo.class::cast)
+        .findFirst()
+        .orElse(null);
+
+    if (ddfInfo == null) {
+      LOGGER.info("Unable to handle any of the metadatas received in this task");
+      return false;
+    }
+
+    return true;
+  }
+
+  private void handle(Task task) throws InterruptedException {
+    NodeAdapter destination = getDestinationAdapter(task);
+    if (destination == null) {
+      task.unlock();
+      return;
+    }
+
+    if (!adaptersAvailable(task, destination)) {
+      return;
+    }
+
+    List<DDFMetadataInfo> infos = task.metadatas()
+        .filter(DDFMetadataInfo.class::isInstance)
+        .map(DDFMetadataInfo.class::cast)
+        .collect(Collectors.toList());
+
+    final DDFMetadataInfo info = infos.get(0);
+    final Metadata metadata = new MetadataImpl(((Map) info.getData()).get("raw-metadata"),
```

Comment:
This will change when https://github.com/connexta/replication/pull/215 goes in. It should simplify this I think.